### PR TITLE
Show error message in tip dialog if one-time tip fails

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/19071

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See issue.

In addition to testing the "tip error" case, the following cases should be regression tested:

* Changing the monthly tip amount for a publisher.
* Clearing the monthly tip amount for a publisher.
* Sending a tip successfully.
* Sending a tip to a non-verified/non-registered publisher (an entry should be added to the pending tips list).

### CODE REVIEW

**Review by emerick — APPROVED** [review]
LGTM

File: browser/ui/webui/brave_tip_ui.cc [hunk 1]
```diff
@@ -60,6 +60,9 @@ class TipMessageHandler : public WebUIMessageHandler,
   void OnRecurringTipRemoved(RewardsService* rewards_service,
                              bool success) override;
 
+  void OnPendingContributionSaved(RewardsService* rewards_service,
```
Comment (szilardszaloki) [inline browser/ui/webui/brave_tip_ui.cc:63]: Nit: I'm fine with `const` by-value parameters, but I think we agreed on omitting `const` for them.

File: components/brave_rewards/resources/tip/lib/host.ts [hunk 2]
```diff
@@ -171,6 +183,11 @@ export function createHost (): Host {
       if (data.result === 0) {
         chrome.send('fetchBalance')
       }
+      checkPendingTipResult(data.result)
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/resources/tip/lib/host.ts:186]: Exactly - that's the tradeoff with this approach.

[the remark this answers is not in the document]
Comment (szilardszaloki) [reply components/brave_rewards/resources/tip/lib/host.ts:186]: I was wondering if we could at least filter out unwanted contribution types via `type` (which is a `RewardsType`).

File: browser/ui/webui/brave_tip_ui.cc [hunk 3]
```diff
@@ -314,6 +324,7 @@ void TipMessageHandler::OnTip(const base::Value::List& args) {
 
   if (recurring && amount <= 0) {
     rewards_service_->RemoveRecurringTip(publisher_key);
+    OnTipCallback(0, ledger::type::Result::LEDGER_OK);
```
Comment (zenparsing) [inline browser/ui/webui/brave_tip_ui.cc:327]: The tip callback was not getting called when we cleared a monthly tip.

File: components/brave_rewards/resources/tip/lib/host.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/resources/tip/lib/host.ts]: Great idea!

**Review by szilardszaloki — APPROVED** [review]

**Pull request merged**